### PR TITLE
SDFSOF-64-PR1: RPC Actions: implement logic and validations

### PR DESCRIPTION
**Pull request by rkharevych**
<details>
  <summary>PR Checklist</summary>

### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `paymentservice.stellar`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
</details>

### What

RPC actions and unit tests:
- notify_interactive_flow_completed
- request_offchain_funds
- request_onchain_funds
- notify_onchain_funds_received
- notify_offchain_funds_received
- notify_offchain_funds_sent

### Why

RPC actions should replace `PATCH /transactions endpoint`

### Known limitations

[TODO or N/A]

### CODE REVIEW

**Comment by rkharevych** [comment]
I will merge this PR and create new one (`fireblocks_integration` -> `softserve-develop`) as soon as https://github.com/stellar/java-stellar-anchor-sdk/pull/969 is approved

File: platform/src/main/java/org/stellar/anchor/platform/action/NotifyOnchainFundsReceivedHandler.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    if (!((request.getAmountIn() == null
+            && request.getAmountOut() == null
+            && request.getAmountFee() == null)
+        || (request.getAmountIn() != null
+            && request.getAmountOut() != null
+            && request.getAmountFee() != null)
+        || (request.getAmountIn() != null
+            && request.getAmountOut() == null
+            && request.getAmountFee() == null))) {
```
Comment (Ifropc) [inline platform/src/main/java/org/stellar/anchor/platform/action/NotifyOnchainFundsReceivedHandler.java:56]: Can you clarify why this constraint was added?

File: platform/src/main/java/org/stellar/anchor/platform/action/RequestOnchainFundsHandler.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      throw new InvalidParamsException("amount_in is required");
+    }
+    if (request.getAmountOut() == null && txn.getAmountOut() == null) {
+      throw new InvalidParamsException("amount_out is required");
+    }
+    if (request.getAmountFee() == null && txn.getAmountFee() == null) {
+      throw new InvalidParamsException("amount_fee is required");
+    }
+
+    if (sep24DepositInfoGenerator instanceof Sep24DepositInfoNoneGenerator) {
```
Comment (Ifropc) [inline platform/src/main/java/org/stellar/anchor/platform/action/RequestOnchainFundsHandler.java:97]: I think you should throw if memo/destination account is provided and info generator is not "None" (cause this fields will be ignored and business should be notified about it)

File: core/src/main/java/org/stellar/anchor/custody/CustodyService.java [hunk 3]
```diff
@@ -33,4 +33,15 @@ public interface CustodyService {
...
+
+  /**
+   * Create custody transaction refund
+   *
+   * @param txnId transaction ID
+   * @param requestBody request body
+   * @return {@link CreateTransactionPaymentResponse} object
+   * @throws AnchorException if error happens
+   */
+  CreateTransactionPaymentResponse createTransactionRefund(String txnId, Object requestBody)
```
Comment (Ifropc) [inline core/src/main/java/org/stellar/anchor/custody/CustodyService.java:45]: Looks like it's out of scope of this PR. I'd remove it for now

File: api-schema/src/main/java/org/stellar/anchor/api/rpc/action/NotifyOffchainFundsReceivedRequest.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+@Data
+@SuperBuilder
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+public class NotifyOffchainFundsReceivedRequest extends RpcActionParamsRequest {
+
+  private Instant fundsReceivedAt;
+  private String externalTransactionId;
+  private String amountIn;
```
Comment (Ifropc) [inline api-schema/src/main/java/org/stellar/anchor/api/rpc/action/NotifyOffchainFundsReceivedRequest.java:17]: It's a bit unclear just from the naming when amount fields are just values and when it's object of value + asset, maybe we should update the naming?

File: api-schema/src/main/java/org/stellar/anchor/api/rpc/action/NotifyOffchainFundsSentRequest.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.experimental.SuperBuilder;
+
+@Data
+@SuperBuilder
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+public class NotifyOffchainFundsSentRequest extends RpcActionParamsRequest {
+
+  private Instant fundsReceivedAt;
```
Comment (Ifropc) [inline api-schema/src/main/java/org/stellar/anchor/api/rpc/action/NotifyOffchainFundsSentRequest.java:15]:
```suggestion
  private Instant fundsSentAt;
```

**Review by Ifropc — COMMENTED** [review]
LGTM except some minor comments.
Current design looks nice!

File: platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+  }
+
+  public abstract ActionMethod getActionType();
+
+  protected abstract SepTransactionStatus getNextStatus(JdbcSepTransaction txn, T request)
+      throws InvalidRequestException, InvalidParamsException;
+
+  protected abstract Set<SepTransactionStatus> getSupportedStatuses(JdbcSepTransaction txn);
+
+  protected abstract Set<String> getSupportedProtocols();
```
[the remark this answers is not in the document]
Comment (rkharevych) [reply platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java:123]: I updated code to use enum, but i removed `getSupportedProtocols` method and moved this validation to `getSupportedStatuses`, because we still will need to duplicate protocol validation when we add support for SEP-31 etc. Also updated exception message

File: api-schema/src/main/java/org/stellar/anchor/api/rpc/action/NotifyInteractiveFlowCompletedRequest.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+@EqualsAndHashCode(callSuper = false)
+public class NotifyInteractiveFlowCompletedRequest extends RpcActionParamsRequest {
+
+  @NotNull private AmountRequest amountIn;
+
+  @NotNull private AmountRequest amountOut;
+
+  @NotNull private AmountRequest amountFee;
+
+  private String amountExpected;
```
[the remark this answers is not in the document]
Comment (rkharevych) [reply api-schema/src/main/java/org/stellar/anchor/api/rpc/action/NotifyInteractiveFlowCompletedRequest.java:21]: Shouldn't amount_expected be sent in these actions? 
image
In code, for example, in SEP-24 `POST /deposit/interactive` endpoint, the amount is validated, but is not saved in DB:
image
image
However, for `SEP-31`, the amount is stored when the transaction is created:
image
image

File: platform/src/main/java/org/stellar/anchor/platform/action/NotifyOnchainFundsReceivedHandler.java (hunk 1 above)
Comment (rkharevych) [reply platform/src/main/java/org/stellar/anchor/platform/action/NotifyOnchainFundsReceivedHandler.java:56]: When Observer or Custody Server receives a notification about onchain payment, it should call Platform to update SEP transaction and provide `amountIn`(for example, when the user sent less or more than expected). In this case `amountOut` and `amountFee` are optional, because Anchor doesn't know exchange rates, however, they should also be updated, so the business will then call `notify_amounts_updated` action to update them. For offchain payments, there is no such logic because the business is responsible for tracking those payments and will provide all 3 amounts together if needed in `notify_offchain_funds_received` action

File: platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    Set<ConstraintViolation<T>> violations = validator.validate(request);
+    if (CollectionUtils.isNotEmpty(violations)) {
+      throw new InvalidParamsException(
+          violations.stream()
+              .map(ConstraintViolation::getMessage)
+              .collect(joining(System.lineSeparator())));
+    }
+  }
+
+  protected boolean isTrustLineConfigured(String account, String asset) {
```
Comment (Ifropc) [inline platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java:137]: Can you move this 2 methods into util class?

**Review by Ifropc — APPROVED** [review]

File: platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java (hunk 8 above)
Comment (rkharevych) [reply platform/src/main/java/org/stellar/anchor/platform/action/ActionHandler.java:137]: `isTrustLineConfigured` will be moved in the next PR
